src/reuse/global_licensing.py

```diff
@@ -0,0 +1,536 @@
+            UnicodeDecodeError: could not decode file as UTF-8.
+            OSError: some other error surrounding I/O.
+            GlobalLicensingParseError: file could not be parsed.
+        """
+
+    @abstractmethod
+    def reuse_info_of(
+        self, path: StrPath
+    ) -> Dict[GlobalPrecedence, List[ReuseInfo]]:
+        """Find the REUSE information of *path* defined in the configuration.
+        The path must be relative to the root of a
+        :class:`reuse.project.Project`.
+
+        The key indicates the precedence type for the subsequent information.
+        """
+
+
+@attrs.define
+class ReuseDep5(GlobalLicensing):
+    """A soft wrapper around :class:`Copyright`."""
+
+    dep5_copyright: Copyright
+
+    @classmethod
+    def from_file(cls, path: StrPath) -> "ReuseDep5":
+        path = Path(path)
+        try:
+            with path.open(encoding="utf-8") as fp:
+                return cls(str(path), Copyright(fp))
+        except UnicodeDecodeError:
+            raise
+        # TODO: Remove ValueError once
+        # <https://salsa.debian.org/python-debian-team/python-debian/-/merge_requests/123>
+        # is closed
+        except (DebianError, ValueError) as error:
+            raise GlobalLicensingParseError(str(error)) from error
+
+    def reuse_info_of(
+        self, path: StrPath
+    ) -> Dict[GlobalPrecedence, List[ReuseInfo]]:
+        path = PurePath(path).as_posix()
+        result = self.dep5_copyright.find_files_paragraph(path)
+
+        if result is None:
+            return {}
+
+        return {
+            GlobalPrecedence.AGGREGATE: [
+                ReuseInfo(
+                    spdx_expressions=set(
+                        map(_LICENSING.parse, [result.license.synopsis])
+                    ),
+                    copyright_lines=set(
+                        map(str.strip, result.copyright.splitlines())
+                    ),
+                    path=path,
+                    source_type=SourceType.DEP5,
+                    # This is hardcoded. It must be a relative path from the
+                    # project root. self.source is not (guaranteed) a relative
+                    # path.
+                    source_path=".reuse/dep5",
+                )
+            ]
+        }
+
+
+@attrs.define
+class AnnotationsItem:
+    """A class that maps to a single [[annotations]] table element in
+    REUSE.toml.
+    """
+
+    TOML_KEYS = {
+        "paths": "path",
+        "copyright_lines": "SPDX-FileCopyrightText",
+        "spdx_expressions": "SPDX-License-Identifier",
+        "precedence": "precedence",
+    }
+
+    paths: Set[str] = attrs.field(
+        converter=_str_to_set, validator=_validate_set_of_str
+    )
+    copyright_lines: Set[str] = attrs.field(
+        converter=_str_to_set, validator=_validate_set_of_str
+    )
+    spdx_expressions: Set[Expression] = attrs.field(
+        converter=_str_to_set_of_expr, validator=_validate_set_of_expr
+    )
```

Comment:
As-is, these validators are also valid for zero-length sets, meaning that the following REUSE.toml would be 'valid':

```toml
version = 1
[[annotations]]
path = []
SPDX-FileCopyrightText = []
SPDX-License-Identifier = []
precedence = "toml"
```

This is _probably_ fine if we don't forbid zero-length sets in the specification. It's just … meaningless, especially if `path` is empty.

I'm going to see about changing this a little to avoid some awkwardness here, but no solution is going to be perfect.

Reply:
- Forbade path from being zero-length.
- At least one of SPDX-FileCopyrightText and SPDX-License-Identifier must now be non-empty (or even defined).